Mathlib/GroupTheory/Perm/Centralizer.lean

```diff
@@ -0,0 +1,2379 @@
+with given cycle type but we resorted to a more algebraic approach.
+
+Given `g : Equiv.Perm α`, the conjugacy class of `g` is the orbit
+of `g` under the action `ConjAct (Equiv.Perm α)`, and we use
+the orbit-stabilizer theorem
+(`MulAction.card_orbit_mul_card_stabilizer_eq_card_group`)
+to reduce the computation to the computation of the centralizer of `g`,
+the subgroup of `Equiv.Perm α` consisting of all permutations
+which commute with `g`. It is accessed here as
+`MulAction.stabilizer (ConjAct (Equiv.Perm α)) g`.
+
+We compute this subgroup as follows.
+
+* If `h : MulAction.stabilizer (ConjAct (Equiv.Perm α)) g`, then the action
+  of `h` by conjugation on `Equiv.Perm α` stabilizes `g.cycleFactorsFinset`.
+  That induces an action of `MulAction.stabilizer (ConjAct (Equiv.Perm α)) g`
+  on `g.cycleFactorsFinset` which is defined via
+  `Equiv.Perm.OnCycleFactors.subMulActionOnCycleFactors `
+
+* This action defines a group morphism `Equiv.Perm.OnCycleFactors.φ g` from
+  `MulAction.stabilizer (ConjAct (Equiv.Perm α)) g`
+  to `Equiv.Perm (g.cycleFactorsFinset)`
+
+* `Equiv.Perm.OnCycleFactors.Iφ_eq_range` shows that the range of `Equiv.Perm.OnCycleFactors.φ g`
+  is the subgroup `Iφ g` of `Equiv.Perm (g.cycleFactorsFinset)`
+  consisting of permutations `τ` which preserve the length of the cycles.
+  This is showed by constructing a right inverse `Equiv.Perm.OnCycleFactors.φ'`
+  in `Equiv.Perm.OnCycleFactors.hφ'_is_rightInverse`.
+
+* `Equiv.Perm.OnCycleFactors.hφ_range_card` computes the cardinality of
+  `range (Equiv.Perm.OnCycleFactors.φ g)` as a product of factorials.
+
+* For an element `z : Equiv.Perm α`, we then prove in
+  `Equiv.Perm.OnCycleFactors.hφ_mem_ker_iff` that `ConjAct.toConjAct z` belongs to
+  the kernel of `Equiv.Perm.OnCycleFactors.φ g` if and only if it permutes `g.fixedPoints`
+  and it acts on each cycle of `g` as a power of that cycle.
+  This gives a description of the kernel of `Equiv.Perm.OnCycleFactors.φ g` as the product
+  of a symmetric group and of a product of cyclic groups.
+  This analysis starts with the morphism `Equiv.Perm.OnCycleFactors.θ`,
+  its injectivity `Equiv.Perm.OnCycleFactors.hθ_injective`,
+  its range `Equiv.Perm.OnCycleFactors.hφ_ker_eq_θ_range`,
+  and  its cardinality `Equiv.Perm.OnCycleFactors.hθ_range_card`.
+
+* `Equiv.Perm.conj_stabilizer_card g` computes the cardinality
+  of the centralizer of `g`
+
+* `Equiv.Perm.conj_class_card_mul_eq g` computes the cardinality
+  of the conjugacy class of `g`.
+
+* We now can compute the cardinality of the set of permutations with given cycle type.
+  The condition for this cardinality to be zero is given by
+  `Equiv.Perm.card_of_cycleType_eq_zero_iff`
+  which is itself derived from `Equiv.Perm.exists_with_cycleType_iff`.
+
+* `Equiv.Perm.card_of_cycleType_mul_eq m` and `Equiv.Perm.card_of_cycleType m`
+  compute this cardinality.
+
+* `AlternatingGroup.of_cycleType_eq m`, `AlternatingGroup.of_cycleType_eq m`
```

Comment:
This line mentions the same name twice... did you mean something else?

Reply:
no, it's a typo.